File: CalCal/Services/NutritionService.swift
import Foundation

// Custom, more descriptive errors
enum NutritionError: Error, LocalizedError {
    case invalidAPIKey
    case badRequest
    case badResponse
    case unparsableJSON(String)
    case apiError(String)

    var errorDescription: String? {
        switch self {
        case .invalidAPIKey:
            return "Invalid OpenRouter API Key. Please check your key in Gemini-Info.plist."
        case .badRequest:
            return "The request to the server was malformed."
        case .badResponse:
            return "The server returned an invalid response."
        case .unparsableJSON(let details):
            return "Could not parse the nutrition data from the AI. Details: \(details)"
        case .apiError(let message):
            return "OpenRouter API Error: \(message)"
        }
    }
}

// MARK: - Networking Service
class NutritionService {
    private let apiKey: String
    private let modelName: String
    private let apiURL = URL(string: "https://openrouter.ai/api/v1/chat/completions")!
    private let serperService = SerperService()
    private let offService = OpenFoodFactsService()

    init() {
        guard let apiKey = APIKeyManager.getOpenRouterAPIKey() else {
            fatalError("OpenRouter API Key not found. Please add it to Gemini-Info.plist")
        }
        self.apiKey = apiKey
        
        guard let modelName = APIKeyManager.getModelName() else {
             fatalError("Model Name not found. Please add MODEL_NAME to Gemini-Info.plist")
        }
        self.modelName = modelName
    }

    func fetchNutrition(for query: String, images: [Data] = []) async throws -> [NutritionResponse] {
        var request = URLRequest(url: apiURL)
        request.httpMethod = "POST"
        request.addValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        // Add HTTP Referer (required by OpenRouter for some tiers, good practice)
        request.addValue("https://calcal.app", forHTTPHeaderField: "HTTP-Referer")
        request.addValue("CalCal", forHTTPHeaderField: "X-Title")

        var systemPrompt = """
        You are a highly accurate nutritional analysis expert.
        Analyze the food query and images provided by the user to identify each distinct food item and return its nutritional information.
        
        If the user query mentions a specific restaurant, brand, or a complex food item that you are not 100% sure about, you MUST use available tools to find the most accurate and up-to-date nutritional information.
        
        TOOL PRIORITY RULE:
        1. `openfoodfacts_search`: Use this FIRST for any branded, packaged, or barcoded product query (e.g., "Nutella", "Chobani", "Oreo"). The data is structured and highly reliable.
           CRITICAL: When using `openfoodfacts_search`, pass ONLY the brand and product name (e.g., "Coke Zero", "Snickers", "Сырок Ростагроэкспорт"). DO NOT include weights, volumes, or packaging details (e.g., "0.33l", "50g", "box") in the search query, as this often causes the search to fail. The tool will return available sizes/quantities for you to select from.
        2. `google_search`: Use this if `openfoodfacts_search` returns no results, or for restaurant menu items ("Big Mac"), generic dishes ("Caesar Salad"), or specific queries requiring web synthesis.
        
        CRITICAL PORTION ESTIMATION RULE: For branded or packaged items, if the user does not specify a weight, you MUST use the serving size or unit weight returned by the tools. NEVER default to 100g if the standard unit weight is different.
        
        CRITICAL SEARCH LANGUAGE RULE: Always perform searches in the SAME language as the user's input query to ensure the most relevant local results.
        
        CRITICAL: If you used a tool to find information for a food item, you MUST set the "isSearchGrounded" key to true for that item in your JSON response.
        
        CRITICAL: Your final response MUST be ONLY a single, minified JSON object with the "foods" array.
        """
        
        var userPrompt = "Analyze the food query: '\(query)'."
        
        if !images.isEmpty {
            userPrompt += " The user has also provided images of the food. Use them to identify the food and estimate portions."
        }
        
        userPrompt += """
        
        The JSON object must have a single key "foods" which is an array of objects. Each object in the array must have these exact keys and value types:
        - "identifiedFood": String (A descriptive name, e.g., "1 large apple")
        - "cleanFoodName": String (A simple, clean name for the food, e.g., "Apple" or "Beef Patty". This should not include quantities or weights.)
        - "calories": Double (for the estimated portion weight)
        - "protein": Double (for the estimated portion weight)
        - "carbs": Double (for the estimated portion weight)
        - "fat": Double (for the estimated portion weight)
        - "estimatedWeightGrams": Double (The realistic weight of the portion. For branded items without weight specified, use the weight of ONE standard package/unit. Use search if unknown.)
        - "caloriesPer100g": Double
        - "proteinPer100g": Double
        - "carbsPer100g": Double
        - "fatPer100g": Double
        - "isSearchGrounded": Boolean
        - "dataSource": String (Optional: "OFF" for OpenFoodFacts, "Google" for Google Search, or null/omitted for internal knowledge)
        If the query is "an apple and a banana", you must return two separate objects in the "foods" array. If the query is "a glass of milk", return one object in the array.
        CRITICAL: The `identifiedFood` and `cleanFoodName` strings in your JSON response MUST be in the same language as the input query.
        """
        
        var contentParts: [OpenRouterRequest.ContentPart] = [.text(userPrompt)]
        
        for imageData in images {
            let base64 = imageData.base64EncodedString()
            let url = "data:image/jpeg;base64,\(base64)"
            contentParts.append(.imageUrl(url))
        }
        
        var messages: [OpenRouterRequest.Message] = [
            .init(role: "system", content: [.text(systemPrompt)]),
            .init(role: "user", content: contentParts)
        ]
        
        let tools: [OpenRouterRequest.Tool] = [
            .init(
                type: "function",
                function: .init(
                    name: "google_search",
                    description: "Search Google for nutritional information or standard portion/unit weights of specific food items, brands, or restaurant menu items.",
                    parameters: [
                        "type": .string("object"),
                        "properties": .object([
                            "query": .object([
                                "type": .string("string"),
                                "description": .string("The search query, e.g., 'McDonalds Big Mac nutrition facts', 'weight of one Snickers bar', or 'Сырок Ростагроэкспорт вес 1 шт'.")
                            ])
                        ]),
                        "required": .array([.string("query")])
                    ]
                )
            ),
            .init(
                type: "function",
                function: .init(
                    name: "openfoodfacts_search",
                    description: "Search OpenFoodFacts database for branded, packaged products to get precise nutritional data and serving sizes.",
                    parameters: [
                        "type": .string("object"),
                        "properties": .object([
                            "query": .object([
                                "type": .string("string"),
                                "description": .string("The product name or brand to search for, e.g., 'Nutella', 'Coca Cola', 'Oreo'.")
                            ])
                        ]),
                        "required": .array([.string("query")])
                    ]
                )
            )
        ]

        var capturedSearchSteps: [SearchStep] = []

        // Loop for tool calling
        for _ in 0...3 { // Limit to 3 iterations to avoid infinite loops
            #if DEBUG
            print("\n--- OPENROUTER REQUEST ---")
            print("Model: \(self.modelName)")
            for (index, msg) in messages.enumerated() {
                print("Message [\(index)] (\(msg.role)):")
                if let content = msg.content {
                    for part in content {
                        switch part {
                        case .text(let text):
                            print("  Text: \(text)")
                        case .imageUrl(let url):
                            if url.hasPrefix("data:image") {
                                print("  Image: [Base64 Data Truncated]")
                            } else {
                                print("  Image URL: \(url)")
                            }
                        }
                    }
                }
                if let toolCalls = msg.toolCalls {
                    for tc in toolCalls {
                        print("  Tool Call: \(tc.function.name)(\(tc.function.arguments))")
                    }
                }
                if let tcId = msg.toolCallId {
                    print("  Tool Call ID: \(tcId)")
                }
            }
            #endif

            let openRouterRequest = OpenRouterRequest(
                model: self.modelName,
                messages: messages,
                responseFormat: nil,
                tools: tools,
                toolChoice: .auto
            )
            
            request.httpBody = try JSONEncoder().encode(openRouterRequest)

            let (data, response) = try await URLSession.shared.data(for: request)
            
            #if DEBUG
            if let responseString = String(data: data, encoding: .utf8) {
                print("\n--- OPENROUTER RESPONSE ---")
                print(responseString)
                print("---------------------------\n")
            }
            #endif
            
            guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
                if let httpResponse = response as? HTTPURLResponse {
                     print("--- NETWORKING ERROR (fetchNutrition) ---")
                     print("Status Code: \(httpResponse.statusCode)")
                     print("Response Body: \(String(data: data, encoding: .utf8) ?? "Unable to print body")")
                     
                     if let errorResponse = try? JSONDecoder().decode(OpenRouterErrorResponse.self, from: data) {
                        throw NutritionError.apiError(errorResponse.error.message)
                     }

                     if httpResponse.statusCode == 401 || httpResponse.statusCode == 403 {
                        throw NutritionError.invalidAPIKey
                     }
                }
                throw NutritionError.badResponse
            }

            let openRouterResponse = try JSONDecoder().decode(OpenRouterResponse.self, from: data)
            guard let choice = openRouterResponse.choices.first else {
                throw NutritionError.badResponse
            }
            
            let message = choice.message
            
            // Add the assistant's message to the history
            messages.append(.init(
                role: "assistant",
                content: message.content != nil ? [.text(message.content!)] : nil,
                toolCalls: message.toolCalls
            ))

            if choice.finishReason == "tool_calls", let toolCalls = message.toolCalls {
                for toolCall in toolCalls {
                    if toolCall.function.name == "google_search" {
                        guard let argsData = toolCall.function.arguments.data(using: .utf8),
                              let args = try? JSONSerialization.jsonObject(with: argsData) as? [String: Any],
                              let searchQuery = args["query"] as? String else {
                            continue
                        }
                        
                        let searchStep = try await serperService.searchStructured(query: searchQuery)
                        capturedSearchSteps.append(searchStep)
                        
                        // Convert back to string for the AI
                        var resultString = ""
                        if let answer = searchStep.answerBox {
                            resultString += "Answer: \(answer)\n"
                        }
                        resultString += "Top results (4):\n"
                        for (i, res) in searchStep.results.enumerated() {
                            resultString += "\(i+1). \(res.title): \(res.snippet)\n"
                        }
                        
                        messages.append(.init(
                            role: "tool",
                            content: [.text(resultString)],
                            toolCallId: toolCall.id,
                            name: "google_search"
                        ))
                    } else if toolCall.function.name == "openfoodfacts_search" {
                        guard let argsData = toolCall.function.arguments.data(using: .utf8),
                              let args = try? JSONSerialization.jsonObject(with: argsData) as? [String: Any],
                              let searchQuery = args["query"] as? String else {
                            continue
                        }
                        
                        var resultString = ""
                        do {
                            let products = try await offService.searchProducts(query: searchQuery)
                            
                            if products.isEmpty {
                                resultString = "No products found in OpenFoodFacts database."
                            } else {
                                resultString = "Found \(products.count) products. Top 3 relevant results:\n"
                                for (i, product) in products.prefix(3).enumerated() {
                                    resultString += "\(i+1). Name: \(product.productName ?? "Unknown") | Brand: \(product.brands ?? "Unknown") | Serving: \(product.servingSize ?? "Unknown")\n"
                                    if let nuts = product.nutriments {
                                        resultString += "   Per 100g: \(nuts.energyKcal100g ?? 0) kcal, P: \(nuts.proteins100g ?? 0)g, C: \(nuts.carbohydrates100g ?? 0)g, F: \(nuts.fat100g ?? 0)g\n"
                                    }
                                }
                            }
                        } catch {
                            resultString = "Error searching OpenFoodFacts: \(error.localizedDescription)"
                        }
                        
                        messages.append(.init(
                            role: "tool",
                            content: [.text(resultString)],
                            toolCallId: toolCall.id,
                            name: "openfoodfacts_search"
                        ))
                    }
                }
                // Continue the loop to get the next response from the LLM
                continue
            } else {
                // Final response
                guard var nutritionJSONText = message.content else {
                    throw NutritionError.badResponse
                }

                if let jsonStartIndex = nutritionJSONText.firstIndex(of: "{"),
                   let jsonEndIndex = nutritionJSONText.lastIndex(of: "}") {
                    nutritionJSONText = String(nutritionJSONText[jsonStartIndex...jsonEndIndex])
                }

                do {
                    let foodArrayResponse = try JSONDecoder().decode(FoodArrayResponse.self, from: Data(nutritionJSONText.utf8))
                    var foods = foodArrayResponse.foods
                    
                    // Attach search steps to any item that is grounded
                    if !capturedSearchSteps.isEmpty {
                        for i in 0..<foods.count {
                            if foods[i].isSearchGrounded == true {
                                foods[i].searchSteps = capturedSearchSteps
                            }
                        }
                    }
                    
                    return foods
                } catch let decodingError {
                    throw NutritionError.unparsableJSON(decodingError.localizedDescription)
                }
            }
        }
        
        throw NutritionError.badResponse
    }
    
    func fetchAIGoals(userStats: String, userGoals: String, baselineTDEE: Double) async throws -> GoalResponse {
         var request = URLRequest(url: apiURL)
        request.httpMethod = "POST"
        request.addValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
        request.addValue("application/json", forHTTPHeaderField: "Content-Type")
        request.addValue("https://calcal.app", forHTTPHeaderField: "HTTP-Referer")
        request.addValue("CalCal", forHTTPHeaderField: "X-Title")

        let prompt = """
        Act as a nutrition planning expert. Based on the following user data, determine their daily nutritional goals.
        User Data: \(userStats)
        User's Personal Goals: "\(userGoals)"
        The user's calculated baseline TDEE (Total Daily Energy Expenditure) for maintenance is \(String(format: "%.0f", baselineTDEE)) calories. Use this as a starting point.

        CRITICAL: Your entire response must be ONLY a single, minified JSON object. Do not include any other text, explanations, or markdown formatting.
        The JSON object must have these exact keys and value types:
        - "calories": Double
        - "protein": Double
        - "carbs": Double
        - "fat": Double
        - "explanation": String
        
        In the "explanation" string, you MUST do the following:
        1.  Start by explaining how you adjusted the baseline TDEE to arrive at the new calorie goal, explicitly referencing the user's personal goal. (e.g., "To help you achieve your goal of losing weight, I've applied a 20% calorie deficit to your baseline TDEE of \(String(format: "%.0f", baselineTDEE)) kcal, resulting in a target of...").
        2.  Briefly explain the macronutrient split (e.g., "This high-protein diet will support muscle growth...").
        3.  Provide a simple, sample one-day meal plan (e.g., Breakfast, Lunch, Dinner, Snacks) with specific food examples that would help the user meet these new targets.
        """

        let openRouterRequest = OpenRouterRequest(
            model: self.modelName,
            messages: [.init(role: "user", content: [.text(prompt)])],
            responseFormat: nil 
        )
        
        #if DEBUG
        print("\n--- OPENROUTER REQUEST (fetchAIGoals) ---")
        print("Model: \(self.modelName)")
        print("Prompt: \(prompt)")
        #endif

        request.httpBody = try JSONEncoder().encode(openRouterRequest)

        let (data, response) = try await URLSession.shared.data(for: request)
        
        #if DEBUG
        if let responseString = String(data: data, encoding: .utf8) {
            print("\n--- OPENROUTER RESPONSE (fetchAIGoals) ---")
            print(responseString)
            print("---------------------------\n")
        }
        #endif
        
        guard let httpResponse = response as? HTTPURLResponse, httpResponse.statusCode == 200 else {
            if let httpResponse = response as? HTTPURLResponse {
                print("--- NETWORKING ERROR (AIGoals) ---")
                print("Status Code: \(httpResponse.statusCode)")
                print("Response Body: \(String(data: data, encoding: .utf8) ?? "Unable to print body")")
                
                if let errorResponse = try? JSONDecoder().decode(OpenRouterErrorResponse.self, from: data) {
                    throw NutritionError.apiError(errorResponse.error.message)
                }

                if httpResponse.statusCode == 401 || httpResponse.statusCode == 403 {
                    throw NutritionError.invalidAPIKey
                }
            }
            throw NutritionError.badResponse
        }

        let openRouterResponse = try JSONDecoder().decode(OpenRouterResponse.self, from: data)
        guard var goalJSONText = openRouterResponse.choices.first?.message.content else {
            throw NutritionError.badResponse
        }

        if let jsonStartIndex = goalJSONText.firstIndex(of: "{"),
           let jsonEndIndex = goalJSONText.lastIndex(of: "}") {
            goalJSONText = String(goalJSONText[jsonStartIndex...jsonEndIndex])
        }

        do {
            return try JSONDecoder().decode(GoalResponse.self, from: Data(goalJSONText.utf8))
        } catch let decodingError {
            throw NutritionError.unparsableJSON(decodingError.localizedDescription)
        }
    }
}

// MARK: - Codable Structs for Domain Models

// This is the new top-level response for food queries
struct FoodArrayResponse: Codable {
    let foods: [NutritionResponse]
}

struct NutritionResponse: Codable {
    let identifiedFood: String
    let cleanFoodName: String
    let calories: Double
    let protein: Double
    let carbs: Double
    let fat: Double
    let estimatedWeightGrams: Double
    let caloriesPer100g: Double
    let proteinPer100g: Double
    let carbsPer100g: Double
    let fatPer100g: Double
    let isSearchGrounded: Bool?
    let dataSource: String?
    var searchSteps: [SearchStep]?
}

struct GoalResponse: Codable {
    let calories: Double
    let protein: Double
    let carbs: Double
    let fat: Double
    let explanation: String
}

File: Android app/app/src/main/java/com/example/calcal/data/remote/NutritionService.kt
package com.example.calcal.data.remote

import android.util.Base64
import com.example.calcal.domain.model.FoodArrayResponse
import com.example.calcal.domain.model.GoalResponse
import com.example.calcal.domain.model.NutritionResponse
import com.example.calcal.domain.model.NutritionResponseWithSteps
import com.example.calcal.domain.model.SearchStep
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withContext
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.*
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class NutritionService @Inject constructor(
    private val okHttpClient: OkHttpClient,
    private val serperService: SerperService,
    private val offService: OpenFoodFactsService
) {
    private val apiUrl = "https://openrouter.ai/api/v1/chat/completions"
    private val json = Json { ignoreUnknownKeys = true; coerceInputValues = true }

    private val toolsJson = buildJsonArray {
        addJsonObject {
            put("type", "function")
            putJsonObject("function") {
                put("name", "google_search")
                put("description", "Search Google for nutritional information or standard portion/unit weights.")
                putJsonObject("parameters") {
                    put("type", "object")
                    putJsonObject("properties") {
                        putJsonObject("query") {
                            put("type", "string")
                            put("description", "The search query")
                        }
                    }
                    putJsonArray("required") { add("query") }
                }
            }
        }
        addJsonObject {
            put("type", "function")
            putJsonObject("function") {
                put("name", "openfoodfacts_search")
                put("description", "Search OpenFoodFacts for branded/packaged products.")
                putJsonObject("parameters") {
                    put("type", "object")
                    putJsonObject("properties") {
                        putJsonObject("query") {
                            put("type", "string")
                            put("description", "Product name or brand")
                        }
                    }
                    putJsonArray("required") { add("query") }
                }
            }
        }
    }

    suspend fun fetchNutrition(
        query: String,
        images: List<ByteArray> = emptyList(),
        apiKey: String,
        modelName: String,
        serperApiKey: String
    ): List<NutritionResponseWithSteps> {
        val initialSystemPrompt = buildInitialSystemPrompt()
        val finalSystemPrompt = buildFinalSystemPrompt()

        val userContentParts = buildJsonArray {
            addJsonObject {
                put("type", "text")
                put("text", "Analyze the food query: '$query'." +
                        if (images.isNotEmpty()) " The user has also provided images." else "")
            }
            images.forEach { imgBytes ->
                val b64 = Base64.encodeToString(imgBytes, Base64.NO_WRAP)
                addJsonObject {
                    put("type", "image_url")
                    putJsonObject("image_url") { put("url", "data:image/jpeg;base64,$b64") }
                }
            }
        }

        var messages = buildJsonArray {
            addJsonObject { put("role", "system"); put("content", initialSystemPrompt) }
            addJsonObject { put("role", "user"); put("content", userContentParts) }
        }

        var capturedSearchSteps = mutableListOf<SearchStep>()
        var didUseOFF = false
        var isAnalysisPass = false

        repeat(4) { // max 4 iterations
            val requestBody = buildJsonObject {
                put("model", modelName)
                put("messages", messages)
                if (!isAnalysisPass) {
                    put("tools", toolsJson)
                    put("tool_choice", "auto")
                }
                putJsonObject("reasoning") { put("effort", "low") }
            }

            val response = postJson(apiUrl, requestBody.toString(), apiKey)
            val responseObj = json.parseToJsonElement(response).jsonObject
            val choice = responseObj["choices"]?.jsonArray?.firstOrNull()?.jsonObject
                ?: throw Exception("No choices in response")

            val message = choice["message"]?.jsonObject ?: throw Exception("No message")
            val finishReason = choice["finish_reason"]?.jsonPrimitive?.contentOrNull

            // Append assistant message to history
            val updatedMessages = messages.toMutableList()
            updatedMessages.add(message)
            messages = JsonArray(updatedMessages)

            if (finishReason == "tool_calls") {
                val toolCalls = message["tool_calls"]?.jsonArray ?: return@repeat

                // Execute tool calls in parallel
                val toolResults = coroutineScope {
                    toolCalls.map { toolCallEl ->
                        async {
                            val toolCall = toolCallEl.jsonObject
                            val toolId = toolCall["id"]?.jsonPrimitive?.content ?: ""
                            val funcObj = toolCall["function"]?.jsonObject ?: return@async null
                            val funcName = funcObj["name"]?.jsonPrimitive?.content ?: ""
                            val argsStr = funcObj["arguments"]?.jsonPrimitive?.content ?: "{}"
                            val args = json.parseToJsonElement(argsStr).jsonObject
                            val q = args["query"]?.jsonPrimitive?.contentOrNull ?: ""

                            when (funcName) {
                                "google_search" -> {
                                    try {
                                        val step = serperService.searchStructured(q, serperApiKey)
                                        capturedSearchSteps.add(step)
                                        var result = ""
                                        step.answerBox?.let { result += "Answer: $it\n" }
                                        result += "Top results:\n"
                                        step.results.forEachIndexed { i, r -> result += "${i+1}. ${r.title}: ${r.snippet}\n" }
                                        Triple(toolId, result, "google")
                                    } catch (e: Exception) {
                                        Triple(toolId, "Error: ${e.message}", "error")
                                    }
                                }
                                "openfoodfacts_search" -> {
                                    try {
                                        val products = offService.searchProducts(q)
                                        didUseOFF = true
                                        val result = if (products.isEmpty()) {
                                            "No products found in OpenFoodFacts."
                                        } else {
                                            buildString {
                                                append("Found ${products.size} products. Top 3:\n")
                                                products.take(3).forEachIndexed { i, p ->
                                                    append("${i+1}. ${p.productName ?: "Unknown"} | ${p.brands ?: ""} | serving: ${p.servingSize ?: "?"}\n")
                                                    p.nutriments?.let { n ->
                                                        append("   Per 100g: ${n.energyKcal100g ?: 0} kcal, P:${n.proteins100g ?: 0}g, C:${n.carbohydrates100g ?: 0}g, F:${n.fat100g ?: 0}g\n")
                                                    }
                                                }
                                            }
                                        }
                                        Triple(toolId, result, "off")
                                    } catch (e: Exception) {
                                        Triple(toolId, "Error: ${e.message}", "error")
                                    }
                                }
                                else -> Triple(toolId, "Unknown tool", "error")
                            }
                        }
                    }.awaitAll().filterNotNull()
                }

                val resultsSummary = toolResults.joinToString("\n") { (_, content, type) ->
                    when (type) {
                        "google" -> "[Search Result]: $content"
                        "off" -> "[Branded Product Data]: $content"
                        else -> "[Tool Error]: $content"
                    }
                }

                val finalUserPrompt = "User's Original Query: \"$query\"\n\nGathered Information:\n$resultsSummary\n\nBased on the information above, provide the final nutritional analysis."

                messages = buildJsonArray {
                    addJsonObject { put("role", "system"); put("content", finalSystemPrompt) }
                    addJsonObject { put("role", "user"); put("content", finalUserPrompt) }
                }
                isAnalysisPass = true

            } else {
                // Final response — parse JSON
                var nutritionText = message["content"]?.jsonPrimitive?.contentOrNull
                    ?: throw Exception("No content in final response")

                val start = nutritionText.indexOf('{')
                val end = nutritionText.lastIndexOf('}')
                if (start >= 0 && end > start) nutritionText = nutritionText.substring(start, end + 1)

                val foodArray = json.decodeFromString<FoodArrayResponse>(nutritionText)
                var foods = foodArray.foods.toMutableList()

                // Attach search grounding info
                if (didUseOFF) {
                    foods = foods.map { food ->
                        val currentSource = food.dataSource
                        val newSource = when {
                            currentSource.isNullOrBlank() -> "OFF"
                            currentSource.contains("OFF") -> currentSource
                            else -> "$currentSource, OFF"
                        }
                        food.copy(
                            isSearchGrounded = true,
                            dataSource = newSource
                        )
                    }.toMutableList()
                } else if (capturedSearchSteps.isNotEmpty()) {
                    foods = foods.map { food ->
                        if (food.isSearchGrounded == true) food
                        else food.copy(isSearchGrounded = true)
                    }.toMutableList()
                }

                return foods.map { food ->
                    NutritionResponseWithSteps(
                        response = food,
                        searchSteps = if (food.isSearchGrounded == true) capturedSearchSteps else emptyList()
                    )
                }
            }
        }

        throw Exception("Max tool-call iterations reached")
    }

    suspend fun fetchAIGoals(
        userStats: String,
        userGoals: String,
        baselineTdee: Double,
        apiKey: String,
        modelName: String
    ): GoalResponse {
        val prompt = """
            Act as a nutrition planning expert. Based on the following user data, determine their daily nutritional goals.
            User Data: $userStats
            User's Personal Goals: "$userGoals"
            The user's calculated baseline TDEE is ${baselineTdee.toInt()} calories.
            
            CRITICAL: Your entire response must be ONLY a single, minified JSON object with keys:
            "calories", "protein", "carbs", "fat" (all Double), "explanation" (String).
        """.trimIndent()

        val requestBody = buildJsonObject {
            put("model", modelName)
            putJsonArray("messages") {
                addJsonObject { put("role", "user"); put("content", prompt) }
            }
        }

        val response = postJson(apiUrl, requestBody.toString(), apiKey)
        val responseObj = json.parseToJsonElement(response).jsonObject
        var goalText = responseObj["choices"]?.jsonArray?.firstOrNull()
            ?.jsonObject?.get("message")?.jsonObject?.get("content")?.jsonPrimitive?.content
            ?: throw Exception("No content in goal response")

        val start = goalText.indexOf('{')
        val end = goalText.lastIndexOf('}')
        if (start >= 0 && end > start) goalText = goalText.substring(start, end + 1)

        return json.decodeFromString(goalText)
    }

    private suspend fun postJson(url: String, body: String, apiKey: String): String = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url(url)
            .post(body.toRequestBody("application/json".toMediaType()))
            .addHeader("Authorization", "Bearer $apiKey")
            .addHeader("Content-Type", "application/json")
            .addHeader("HTTP-Referer", "https://calcal.app")
            .addHeader("X-Title", "CalCal")
            .build()

        val response = okHttpClient.newCall(request).execute()
        val responseBody = response.body?.string() ?: throw Exception("Empty response")
        if (!response.isSuccessful) throw Exception("API error ${response.code}: $responseBody")
        responseBody
    }

    private fun buildInitialSystemPrompt() = """
        You are a highly accurate nutritional analysis expert.
        Analyze the food query and images provided by the user to identify each distinct food item.
        
        LANGUAGE & BRAND RULE: The query can be in ANY language (Russian, English, etc.). Brand names may be local/regional brands from any country — do NOT assume a foreign-sounding name maps to a well-known global brand. For example, "актимуно" is a Russian brand and is NOT the same as "Actimel". Always search for the exact name as given.
        
        CRITICAL SEARCH RULE: You MUST use tools for ANY of the following — no exceptions:
        - Any branded or packaged product (Oreo, Activia, Lay's, Snickers, Актимуно, etc.)
        - Any product with a recognizable brand name, even if you think you know the nutrition
        - Any restaurant or fast food item
        - Any product name that sounds like a brand (even if unfamiliar or in a foreign language)
        - Any query where the user specifies a quantity of a packaged item (e.g. "3 oreo", "2 актимуно")
        - When in doubt — always search, never guess
        
        Only skip tools for completely generic, unbranded whole foods (e.g. "apple", "boiled egg", "rice", "яблоко", "варёное яйцо").
        
        CRITICAL TOOL BATCHING RULE: Emit ALL necessary tool calls in a SINGLE response turn. Execute them in parallel.
        
        TOOL PRIORITY RULE:
        1. openfoodfacts_search: Use FIRST for ALL branded/packaged products. Pass ONLY brand and product name, no weights or quantities.
        2. google_search: Use if OFF returns no results, or for restaurant items/generic dishes.
        
        Do not output JSON yet. Focus on gathering data.
    """.trimIndent()

    private fun buildFinalSystemPrompt() = """
        You are a highly accurate nutritional analysis expert.
        
        Analyze the tool results and original query. For branded items without specified weight, use the standard serving size from tools.
        
        If you used a tool, set "isSearchGrounded" to true for that item.
        Set "dataSource" to "OFF" if data came from OpenFoodFacts, "Google" if from Google Search, or null if from internal knowledge.
        
        Your final response MUST be ONLY a single minified JSON object:
        {"foods": [{"identifiedFood": String, "cleanFoodName": String, "calories": Double, "protein": Double, "carbs": Double, "fat": Double, "estimatedWeightGrams": Double, "caloriesPer100g": Double, "proteinPer100g": Double, "carbsPer100g": Double, "fatPer100g": Double, "isSearchGrounded": Boolean, "dataSource": String|null}]}
        
        identifiedFood and cleanFoodName MUST be in the same language as the input query.
    """.trimIndent()
}
